### ISSUE TITLE
Efficient row by row join equivalent of pd.concat axis=1

### ISSUE THREAD

**Issue by daviskirk** [duckdb/duckdb#3423]
I'm looking for a fast way to simply combine the columns of multiple tables/parquet files/pandas dataframes and I'm not sure I'm not overlooking something in duckdb.

Here's an example:
We simply have two tables `df1` and `df2` that we would like to write into a single table (the row indices don't matter, we guarantee that they have the same number of rows).

```python
import duckdb
import numpy as np
import pandas as pd
n, m = 10_000, 100
df1=pd.DataFrame(np.random.random((n, m)), columns=[f"df1_{i}" for i in range(m)]).reset_index()
df2=pd.DataFrame(np.random.random((n, m)), columns=[f"df2_{i}" for i in range(m)]).reset_index()
ddb = duckdb.connect()
```

Using python and pandas:

```python
In [42]: %timeit pd.concat([df1, df2], axis=1, join="inner")
28.7 ms ± 495 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Using duckdb the way I came up with is a simple join:

```python
In [43]: %timeit ddb.execute("select * from df1 inner join df2 on df1.index=df2.index")
40.9 ms ± 1.54 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

I'm assuming this is slower because we're potentially allowing for a lot more complex cases to be handled in the joining:
```
┌───────────────────────────┐                             
│         HASH_JOIN         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│            LEFT           ├──────────────┐              
│        index=index        │              │              
└─────────────┬─────────────┘              │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│        PANDAS_SCAN        ││        PANDAS_SCAN        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           index           ││           index           │
│           df1_0           ││           df2_0           │
│           df1_1           ││           df2_1           │
```

Is there any way to simply combine the columns efficiently completely ignoring any sort of joins or column values?
Or is this basically the correct way to do this sort of simple combining of tables?

**Comment by arjenpdevries** [comment]
I think DuckDB does not have a positional join implementation yet - I agree that this would be a desirable new feature!

**Comment by hawkfish** [comment]
This sounds like an outer join on rowid? If so, we should look at making that fast.

**Comment by danz68** [comment]
This is the feature I miss the most for columnar databases. As a programmer, I am not aware of any language complaining about writting:
for(int i=0; i<n;i++) do_something_with(a[i], b[i]);
How a, b were generated, sorted, resized is nobody concern. Associating two result columns coming from different computations should be allowed. Positional join, as a language extension should be a cheap way of associating data resulting from diferrent computations. 

select a.x, a.y, b.z from a positional join b over (ordered by a.rowid, b.rowid !! is not the same as regular join) would be ideal. In case a and b records count differ, the join should fail. The missing over () contents should instruct the engine to use current processing order to align the resulting values. 

Or at least teach me how to do on tables: 
update a.field = b.field from a, b; 
with no explicit match, just natural/storage order.

**Comment by hawkfish** [comment]
@danz68 Yeah positional join is a common column store physical operator, and the last one I built had it. We are just discussing how to make it parallel so it is faster!

**Comment by hawkfish** [comment]
@danz68 I started to have a look at this and I have to say that the performance is only about ⅓ slower (40ms vs 30ms), which is hardly surprising since we are copying 200 columns into `DataChunk`s and back. I'm not convinced this warrants a separate physical  join operator, but it might if the tables were longer (e.g., 10MR instead of 10KR). I will check the performance at scale some more, but I suspect that this is only worth doing if the hash table gets large.

In any case, I think the way to express most of the operations you are asking about is to join on `rowid`. This is just the syntax SQL uses to express these kinds of operations. It isn't really any different from the `i` index in the procedural example you gave. The loop example would be:

```sql
SELECT do_something(a.*, b.*)
FROM a, b
WHERE a.rowid = b.rowid
```
and the update example would be:

```sql
UPDATE a.field = b.field
FROM a, b
WHERE a.rowid = b.rowid
```

I'm not clear what you meant by:

> select a.x, a.y, b.z from a positional join b over (ordered by a.rowid, b.rowid !! is not the same as regular join) would be ideal. In case a and b records count differ, the join should fail. The missing over () contents should instruct the engine to use current processing order to align the resulting values.

SQL would express this as a join, again on `rowid`, no `OVER` clause required. If you want to detect missing values, you can use an outer join that will insert `NULL`s into the join result - including `rowid` so you could distinguish missing matches from input  `NULL`s.

I think the fundamental conceptual issue here is that SQL operates on _sets_ but Python operates on _arrays_. The set model allows us to do fun things like independent parallel processing of data and out-of-memory processing, but it also means that you have to be explicit about ordering semantics. This may be syntactically onerous at times, but it is just a trade-off necessitated by a different way of thinking about data.  Still, if you can clarify the request a bit I'll think about it some more.

**Comment by hawkfish** [comment]
(I can see using `POSITIONAL JOIN` as syntactic sugar for `rowid = rowid` - that would fit in with our "friendly SQL" mantra. Is that what you meant?)

**Comment by danz68** [comment]
Right now you have implemented loop, merge and hash joins. Each of them with its own use case. A positional join, would be just a syntactic sugar for t1.crt_posid=t2.crt_posid (not the same with rowid, which depends on order and possible deleted records).

Internally would be
for (i=0;i<n;i++) get t1.a[i], t2.b[i]. 
(no specific joining condition, just use the aligned position of current t1 and t2 order)

**Comment by hawkfish** [comment]
What is the difference between `rowid` and `crt_posid`?

**Comment by danz68** [comment]
_"In the current storage, these identifiers are contiguous unsigned integers (0, 1, …) if no rows were deleted. Deletions introduce gaps in the rowids which may be reclaimed later. Therefore, it is strongly recommended not to use rowids as identifiers."_

if rowid is storage related, 
select rowid, f from t 
returns the storage order
0, "B"
1, "A"

select rowid, f from t order by f
(probably) returns
1, "A"
0, "B"

select crt_posid(), f from t
should return
0, "B"
1, "A"

select crt_posid(), f from t order by f
should return
0, "A"
1, "B"

I don't need any form of internal join, a simplified rowid storage join would look like
for (i=0;i<t1.n;i++) {
  a_t1 = t1.a[i];
  ri = t1.rowid[i];
  if ( t2.findbyrowid(ri, &j) ) 
    a_t2 = t2.a[j]; 
  else 
    a_t2 = nul;
}

from my perspective, a positional join would look like
for (i=0;i<t1.n;i++) {
 a_t1 = t1.a[i];
 a_t2 = i<t2.n ? t2.a[i]: nil;
}

which, of course, can be also written as

if (t1.n<=t2.n)
 for (i=0;i<t1.n;i++) {
 a_t1 = t1.a[i];
 a_t2 = t2.a[i];
}
else
{
for (i=0;i<t2.n;i++) {
 a_t1 = t1.a[i];
 a_t2 = t2.a[i];
}
for (i=t2.n;i<t1.n;i++) {
 a_t1 = t1.a[i];
 a_t2 = nil;
}
}

**Comment by arjenpdevries** [comment]
A `positional join` physical operator would be a very valuable addition.

In a preliminary exploration with the JSON extension, I have worked around the lack of such a join by enforcing an index on `rowid`, such that at least an index join is selected. Yet, this is impractical for at least three reasons: the index join would be more expensive than the positional equivalent, we do not really need the index, and, for now, the indexes are not persistent so have to be recreated every time you start your program.

**Comment by hawkfish** [comment]
This is "raw and wriggling", but the perf improvement on the benchmark is about 15x.

**Comment by danz68** [comment]
Ideally, if t1.f is identical with t2.f,
select a,b,c,..,f from t1;
should be as fast as
select t1.a, t1.b, t1.c, ... , **t2.f** from t1 positional join t2 on t1.rowid=t2.rowid;

**Comment by hawkfish** [comment]
> Ideally, if t1.f is identical with t2.f,
> select a,b,c,..,f from t1;
> should be as fast as
> select t1.a, t1.b, t1.c, ... , **t2.f** from t1 positional join t2 on t1.rowid=t2.rowid;

Easy enough to benchmark.

We are closing in on a spec.  Explicit use of `POSITIONAL JOIN` will have the requested semantics and throw if there is a row count mismatch. It will work with `ORDER BY` subqueries as well. 

We will also have the keyhole optimisation I have been testing for the cases where we can determine it applies (ordered, dense, unique keys with a shared base). This includes the basic `rowid` and any other join keys we can determine this for. This optimisation will also support outer join semantics for mismatched table sizes.

**Comment by hawkfish** [comment]
About to take another pass at this, so I'm going to throw in some notes.

The random access table function we need is probably something like
```cpp
typedef void table_function_init_positional_scan_t(ClientContext &context, const FunctionData *bind_data,
                                                  LocalTableFunctionState *local_state,
                                                  GlobalTableFunctionState *global_state, idx_t start_row, idx_t end_row);
```

A second function that hints at row group boundaries would probably help.

But for starters we will just make the single-threaded PJ operator and get that working.

**Comment by hawkfish** [comment]
The new materialised right side straw man version is about 2x faster than the inner join. That should improve when we can avoid materialisation.

**Comment by hawkfish** [comment]
Second dual scan implementation is about 6x faster.
![pj-perf](https://user-images.githubusercontent.com/13156216/210657956-a82698d8-1469-4665-8856-efc5e761e0df.png)

**Comment by danz68** [comment]
Lovely!

I don't know the internals, but I truly believe that "scanning" term is more appropriate than "join" for such operation. In fact what a positional alignment does is a simultaneous chunk retrieval from two or more independent scanning streams, each in its own (programmer) logic order.

It might not be standard SQL, but is standard programming vector scanning.

FETCH
select a.x, a.y, ... from a order by a.whatever,
select b.u, b.v, ... from b order by b.something,
....

would be a very convenient surrogate syntax for an unsupervised data retrieval (no join check).

IMO, equal counters condition is not even needed, the scan length should take the largest one, filling with nulls the shorter streams.

**Comment by hawkfish** [comment]
I believe your example will be covered by the non-scan implementation, which I have not removed. Right now, the planner looks at the two inputs to the logical PJ and uses a scan when they are both table scans and materialisation when they are not. So your example would become:

```sql
SELECT * FROM
    (select a.x, a.y, ... from a order by a.whatever)
    POSITIONAL JOIN
    (select b.u, b.v, ... from b order by b.something)
```

Your comment about making it "outer" by default is interesting because it allows the user the option of filtering out the `NULL`s as the semantics are much simpler than for other joins.

**Comment by danz68** [comment]
Your syntax is clearly SQL-ish, clean and easy. Which I don't mind at all. As long as internally is no actual join performed (hash or alike) to influence the fetching performance any proposed syntax is totally fine. 

As mentioned before, ideally, positional join should not perform any joining test (or seek) which would make it as fast as retrieving data from a single table if it would contain all selected columns from two or more positionally joined tables.

**Comment by hawkfish** [comment]
That's what it does for table scans. For subqueries like your `FETCH` example, it has to materialize one side right now, but we may be able to fiddle with the pipelining infrastructure at some point. The issue is just that pipelines can only have one source. But the table would be materialised anyway if it was sorted, so I don't think that would buy you much in general.

**Comment by danz68** [comment]
I can't wait to try it out in a future release. 

I really hope that at some point DuckDB will be able to use the positional mechanism for updates, too. This would mimic an instant full column replacement mechanism for cases where computed values are not necessarily obtained during a single scan of a table (row expressions). 

It might be too much for a db engine to be able to reproduce any array/matrix operations, yet I admit is very appealing to use a well balanced disk/memory columnar storage for more than basic aggregations or online algorithms.

**Comment by cottrell** [comment]
> Second dual scan implementation is about 6x faster. ![pj-perf](https://user-images.githubusercontent.com/13156216/210657956-a82698d8-1469-4665-8856-efc5e761e0df.png)

If I understand the goal here, the positional scan should actually be O(1) vs rest O(something(n)) ... ignoring read io etc just talking about erasing the join. It literally opens and reads sequentially.

**Comment by danz68** [comment]
POSITIONAL JOIN behaves slower than expected for sequential table scans, a singleton select joining on rowid being a way faster alternative.

I have a 6 mil table (22 columns) imported in a duckdb database. I made another table which contains only one column from the original table, just to check the positional join functionality.

The following two statements will retrieve the same result:

--statement1
select TABLE1.*, TABLE2.* from TABLE1 positional join TABLE2;
--alternative
select * from TABLE1 positional join TABLE2;

--statement2
select TABLE1.*, select (COLUMN_X from TABLE2 where TABLE1.rowid=TABLE2.rowid) from TABLE1;
--alternative
select TABLE1.*, TABLE2 .COLUMN_X from TABLE1 JOIN TABLE2 on TABLE2.rowid=TABLE1.rowid;

but statement2 is way faster than statement 1.

profiling them 
statement1: 1.67s, statement2: 0.62s (2.69x)

I thought that for some kind of reason that the difference would come from fetching top/bottom data so I decided to change a bit the statements, to force a full scan of the results as following:
select distinct * from (select TABLE1.*, TABLE2.* from TABLE1 positional join TABLE2);
select distinct * from (select TABLE1.*, select (COLUMN_X from TABLE2 where TABLE1.rowid=TABLE2.rowid) from TABLE1);

profiling them 
statement1: 11.30s, statement2: 4.86s (2.32x)

========================
Is this behavior expected? Am I not using properly the POSITIONAL JOIN syntax to reach maximum speed for table scans?
(using v0.8.2-dev2673 fd683b9fb2)
